### PR TITLE
Performance and Hash Improvements from Evgeniy

### PR DESCRIPTION
**Pull request by skilesare**
Updates to Hash and EQ functions for Maps, Sets, Bytes, Arrays, Ints, Nats, and Floats.
Fix float equivalence

### CODE REVIEW

File: src/types.mo [hunk 1]
```diff
@@ -481,43 +481,46 @@ module {
...
-          Nat32.toNat(Blob.hash(Blob.fromArray(Buffer.toArray(accumulator))));
+          Nat32.toNat(accumulator);
         };
         case(#Nats(val)){
-          var accumulator = Buffer.Buffer<Nat8>(1);
+          var accumulator = 0 : Nat32;
           for(thisItem in StableBuffer.vals(val)){
-            accumulator.append(Buffer.fromArray<Nat8>(nat32ToBytes(Nat32.fromNat(hash(#Nat(thisItem))))));
+            let ahash = Nat32.fromNat(hash(#Nat(thisItem)));
+            accumulator := (accumulator * 3) +% ahash;
```
Comment (ZhenyaUsenko) [inline src/types.mo]: (accumulator *% 3)

File: src/types.mo [hunk 2]
```diff
@@ -642,20 +659,25 @@ module {
         return true;
       };
       case(#Map(x), #Map(y)){
```
Comment (ZhenyaUsenko) [inline src/types.mo:661]: Why did you change Map as Set comparison? if I am not mistaken, the logic remains the same (does not take order into account), but it now uses sort which is O(n**2) for motoko algorithm if I remember correctly

Comment (skilesare) [reply src/types.mo:661]: Previously the algo would take insertion order into account.  I put sort in so I could compare the items one for one.  I guess I could create a map not use sort.   I could also use get to make sure the item I'm inspecting in x is in y, but I think you said I shouldn't use get.

Comment (ZhenyaUsenko) [reply src/types.mo:661]: I am a bit confused. You was previously looping through mapX and getting equivalent values from mapY to compare (meaning they could be at any position in mapY - not order dependent). Now you pre-sort 2 maps and compare 2 running iterators, which, as far as I understand, concludes to the same overall logic

Comment (skilesare) [reply src/types.mo:661]: Previously you said to not use GET, so I was trying something different, if you think looping through x and Map.getting y is better I will revert.  I think that is better than sorting.

Comment (ZhenyaUsenko) [reply src/types.mo:661]: Yep, if we are talking about algorithm that is NOT order dependent, getting will definitely be better than sorting, sorry for confusing you previously

I was suggesting moving away from get only in a context of an algorithm that IS order dependent
